designs/2025-rule-tester-assertion-options/README.md

```diff
@@ -0,0 +1,391 @@
+
+In most eslint(-plugin)s' rules the error assertions are different from each other.
+Adding options that could be set/shared when configuring the rule tester would ensure a common base level of assertions is met throughout the (plugin's) project.
+The options could be defined on two levels. On `RuleTester`'s `constructor` effectively impacting all tests, or on the test `run` method itself, so only that set is affected.
+
+## Detailed Design
+
+### Variant 1 - Constructor based options
+
+````ts
+new RuleTester(testerConfig: {...}, assertionOptions: {
+    /**
+     * Require message assertion for each invalid test case.
+     * If `true`, errors must extend `string | Array<{ message: string } | { messageId: string }>`.
+     * If `'message'`, errors must extend `string | Array<{ message: string }>`.
+     * If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+     * 
+     * @default false
+     */
+    requireMessage: boolean | 'message' | 'messageId';
+    /**
+     * Require full location assertions for each invalid test case.
+     * If `true`, errors must extend `Array<{ line: number, column: number, endLine: number, endColumn: number }>`.
+     * 
+     * @default false
+     */
+    requireLocation: boolean;
+} = {});
+````
+
+### Variant 2 - Test method based options
+
+````ts
+ruleTester.run("rule-name", rule, tests, assertionOptions: {
+    /**
+     * Require message assertion for each invalid test case.
+     * If `true`, errors must extend `string | Array<{ message: string } | { messageId: string }>`.
+     * If `'message'`, errors must extend `string | Array<{ message: string }>`.
+     * If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+     * 
+     * @default false
+     */
+    requireMessage: boolean | 'message' | 'messageId';
+    /**
+     * Require full location assertions for each invalid test case.
+     * If `true`, errors must extend `Array<{ line: number, column: number, endLine: number, endColumn: number }>`.
+     * 
+     * @default false
+     */
+    requireLocation: boolean;
+    /**
+     * Require and expect only the given test scenarios.
+     * This allows omitting certain scenarios from this run with the current options.
+     * 
+     * @default ["valid","invalid"]
+     */
+    requiredScenarios: ReadonlyArray<'valid' | 'invalid'>;
+});
+````
+
+### Shared Logic
+
+#### requireMessage
+
+If `requireMessage` is set to `true`, the invalid test case cannot consist of an error count assertion only, but must also include a message assertion. (See below)
+This can be done either by providing only a `string` message, or by using the `message`/`messageId` property of the error object in the `TestCaseError` (Same as the current behavior).
+If `true`, errors must extend `string | Array<{ message: string } | { messageId: string }>`.
+If `'message'`, errors must extend `string | Array<{ message: string }>`.
+If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
```

Comment:
Test message can also be a regex. For example, these could be tests for the core `no-undef` rule:

```js
{
	code: "a = 1;",
	errors: [/not defined/],
},
{
	code: "a = 1;",
	errors: [
		{
			message: /not defined/,
		},
	],
},
```

I think that `requireMessage: 'message'` should allow regexes as well, but open to other opinions.